server/command.go

```diff
@@ -320,3 +339,125 @@ func (p *Plugin) executeDisconnectBotCommand(args *model.CommandArgs) (*model.Co
 func getAutocompletePath(path string) string {
 	return "plugins/" + pluginID + "/autocomplete/" + path
 }
+
+func (p *Plugin) getMSTeamsUsersAndMap(client msteams.Client, teamID, channeID string) ([]msteams.User, map[string]msteams.User) {
+	msChannelMembers, err := client.ListChannelMembers(teamID, channeID)
+	if err != nil {
+		p.API.LogError("Unable to get MS Teams users to add in the channel", "error", err)
+		return nil, nil
+	}
+
+	msUsersMap := make(map[string]msteams.User, len(msChannelMembers))
+	for _, u := range msChannelMembers {
+		msUsersMap[u.Mail] = u
+	}
+
+	return msChannelMembers, msUsersMap
+}
+
+func (p *Plugin) getMattermostUsersAndMap(channelID string) ([]*model.User, map[string]*model.User) {
+	mmChannelMembers, getErr := p.API.GetChannelMembers(channelID, 0, math.MaxInt32)
+	if getErr != nil {
+		p.API.LogError("Unable to get Mattermost users to add in the channel", "error", getErr)
+		return nil, nil
+	}
+
+	var mmUsers []*model.User
+	mmUsersMap := make(map[string]*model.User, len(mmChannelMembers))
+	for _, u := range mmChannelMembers {
+		user, err := p.API.GetUser(u.UserId)
+		if err != nil {
+			p.API.LogError("Unable to get the user", "error", err)
+			continue
+		}
+
+		mmUsers = append(mmUsers, user)
+		mmUsersMap[user.Email] = user
+	}
+
+	return mmUsers, mmUsersMap
+}
+
+func (p *Plugin) addMattermostUser(wg *sync.WaitGroup, teamID, channelID string, msChannelMembers []msteams.User, mmUsersMap map[string]*model.User, batchSize int) {
+	for i := 0; i < (len(msChannelMembers)/batchSize)+1; i++ {
+		wg.Add(1)
+		go func(i int) {
```

Comment:
We are creating here an unbounded number of goroutines. Not necessarily a lot, but still, something without any control here. A lot of goroutines hammering the database without control. We have to find a better way to improve that. Probably something that give us the list of users that needs to be created, the list of users that are not in the team, and the list of users that are in the team but not in the channel. And then do the operations on that users, instead of hammering the API all the time with unneeded requests.